### PR TITLE
ARROW-15800 [R] Implement bindings for `lubridate::as_date()` and `lubridate::as_datetime()`

### PR DESCRIPTION
**Pull request by dragosmg**

### CODE REVIEW

File: r/R/dplyr-funcs-type.R [hunk 1]
```diff
@@ -114,14 +113,74 @@ register_bindings_type_cast <- function() {
...
+    if (call_binding("is.numeric", x)) {
+      delta <- call_binding("difftime", origin, "1970-01-01")
+      delta <- build_expr("cast", delta, options = cast_options(to_type = int64()))
+      x <- build_expr("cast", x, options = cast_options(to_type = int64()))
+      output <- build_expr("+", x, delta)
+      output <- build_expr("cast", output, options = cast_options(to_type = timestamp()))
+    } else {
+      output <- build_expr("cast", x, options = cast_options(to_type = timestamp()))
+    }
+    build_expr("assume_timezone", output, options = list(timezone = tz))
```
Comment (dragosmg) [inline r/R/dplyr-funcs-type.R:181]: [Jon's comment](https://github.com/apache/arrow/pull/12677#discussion_r836783740) on the original PR: 
> I can't suggest an edit here for some reason, but could we not rename to output here and then have one line that does build_expr("cast", x, options = cast_options(to_type = timestamp())) for both? (which could be outside of the if|else entirely, and we wouldn't even need an else at all?)

File: r/tests/testthat/test-dplyr-funcs-type.R [hunk 2]
```diff
@@ -875,6 +877,134 @@ test_that("as.Date() converts successfully from date, timestamp, integer, char a
...
+    posixct_var = as.POSIXct("2022-02-25 00:00:01", tz = "Pacific/Marquesas"),
+    date_var = as.Date("2022-02-25"),
+    difference_date = ymd_hms("2010-08-03 00:50:50", tz = "Pacific/Marquesas"),
+    character_ymd_var = "2022-02-25 00:00:01",
+    character_ydm_var = "2022/25/02 00:00:01",
+    integer_var = 32L,
+    integerish_var = 32,
+    double_var = 34.56
+  )
+
```
Comment (dragosmg) [inline r/tests/testthat/test-dplyr-funcs-type.R:891]: [Jon's comment](https://github.com/apache/arrow/pull/12677#discussion_r836786450) on the original PR:
> This fixture looks very similar (though subtly different in some ways) as the one for as.Date() would it make sense to combine them into a single test_that block so they can use one fixture and we don't have to redefine it + make sure we update it in both places, etc?

>If that's not feasible, should we move the fixture out to helpers-data.R so that it's reusable?

File: r/R/dplyr-funcs-type.R [hunk 3]
```diff
@@ -114,14 +113,74 @@ register_bindings_type_cast <- function() {
...
+    } else if (call_binding("is.numeric", x) & !call_binding("is.integer", x)) {
+      # Arrow does not support direct casting from double to date32(), but for
+      # integer-like values we can go via int32()
       # https://issues.apache.org/jira/browse/ARROW-15798
       # TODO revisit if arrow decides to support double -> date casting
-      abort("`as.Date()` with double/float is not supported in Arrow")
+      x <- build_expr("cast", x, options = cast_options(to_type = int32()))
     }
     build_expr("cast", x, options = cast_options(to_type = date32()))
   })
```
Comment (dragosmg) [inline r/R/dplyr-funcs-type.R:168]: [Jon's comment](https://github.com/apache/arrow/pull/12677#discussion_r836782444) on the[ original PR](https://github.com/apache/arrow/pull/12677):
>It looks like a lot of this functionality is duplicated in as.Date up above, could we factor those out into an abstract function that does both?

File: r/R/dplyr-funcs-type.R (hunk 1 above)
Comment (dragosmg) [reply r/R/dplyr-funcs-type.R:181]: Done

File: r/tests/testthat/test-dplyr-funcs-type.R (hunk 2 above)
Comment (dragosmg) [reply r/tests/testthat/test-dplyr-funcs-type.R]: Done. I combined the tests for `as.Date()` and `as_date()` into a single `test_that` block.

File: r/R/dplyr-funcs-type.R [hunk 4]
```diff
@@ -82,44 +82,40 @@ register_bindings_type_cast <- function() {
...
 
-    # cast from numeric
-    } else if (call_binding("is.numeric", x) & !call_binding("is.integer", x)) {
-      # Arrow does not support direct casting from double to date32()
-      # https://issues.apache.org/jira/browse/ARROW-15798
-      # TODO revisit if arrow decides to support double -> date casting
-      abort("`as.Date()` with double/float is not supported in Arrow")
+  register_binding("as_datetime", function(x,
+                                           origin = "1970-01-01",
+                                           tz = "UTC") {
```
Comment (AlenkaF) [inline r/R/dplyr-funcs-type.R:110]: I would think `as_date` accepts any `tz` and `as_datetime` has a default `tz` set to UTC?

Also, how come `as_datetime` doesn't accept `format` - I am wondering how come you did not use `binding_as_date` for `as_datetime` also?

File: r/R/dplyr-funcs-datetime.R [hunk 5]
```diff
@@ -263,11 +263,11 @@ register_bindings_duration <- function() {
     # cast to timestamp if time1 and time2 are not dates or timestamp expressions
     # (the subtraction of which would output a `duration`)
     if (!call_binding("is.instant", time1)) {
-      time1 <- build_expr("cast", time1, options = cast_options(to_type = timestamp(timezone = "UTC")))
+      time1 <- build_expr("cast", time1, options = cast_options(to_type = timestamp()))
     }
 
     if (!call_binding("is.instant", time2)) {
-      time2 <- build_expr("cast", time2, options = cast_options(to_type = timestamp(timezone = "UTC")))
+      time2 <- build_expr("cast", time2, options = cast_options(to_type = timestamp()))
```
Comment (AlenkaF) [inline r/R/dplyr-funcs-datetime.R:270]: Are this changes necessary for this PR or is this a part of something else?

**Review by AlenkaF — COMMENTED** [review]
Looks good! Some comments added to clarify things for me. Thanks!

File: r/R/dplyr-funcs-datetime.R (hunk 5 above)
Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R:270]: Necessary for this PR.

File: r/R/dplyr-funcs-type.R (hunk 4 above)
Comment (dragosmg) [reply r/R/dplyr-funcs-type.R:110]: * I would think `as_date` accepts any `tz` and `as_datetime` has a default tz set to UTC?
**Correct**: the default value for `tz` in `as_datetime()` is `"UTC"` & `as_date()` accepts any `tz`. I changed the default value for `tz` in `as_date()` to `NULL`.

* How come `as_datetime()` doesn't accept `format`?
Oversight, I added support for `format`

* Why I didn't use `binding_as_date()` for `as_datetime`? 
`x` for `as_datetime()` might have a `time` component and `binding_as_date()` won't parse it.

Comment (dragosmg) [reply r/R/dplyr-funcs-type.R]: Done

**Comment by dragosmg** [comment]
@AlenkaF I think this is ready for another look. Thanks!

**Review by AlenkaF — APPROVED** [review]
LGTM, thank you for the changes!

**Comment by dragosmg** [comment]
Thanks @AlenkaF. @thisisnic @jonkeane @nealrichardson would you be able to have a look and merge if everything is ok?

File: r/R/dplyr-funcs-datetime.R [hunk 6]
```diff
@@ -372,3 +372,50 @@ binding_format_datetime <- function(x, format = "", tz = "", usetz = FALSE) {
...
+    # cast from POSIXct
+  } else if (call_binding("is.POSIXct", x)) {
+    # base::as.Date() first converts to the desired timezone and then extracts
+    # the date, which is why we need to go through timestamp() first
+    if (base || !is.null(tz)) {
+      x <- build_expr("cast", x, options = cast_options(to_type = timestamp(timezone = tz)))
+    }
+    # POSIXct is of type double -> we need this to prevent going down the
+    # "double" branch
+    x <- x
```
Comment (jonkeane) [inline r/R/dplyr-funcs-datetime.R:399]: I'm not sure I follow what happens if we don't have this self assignment? This shouldn't be needed (and honestly shouldn't have any effect)

File: r/R/dplyr-funcs-datetime.R [hunk 7]
```diff
@@ -372,3 +372,50 @@ binding_format_datetime <- function(x, format = "", tz = "", usetz = FALSE) {
...
+             (!call_binding("is.integer", x) | origin != "1970-01-01")) {
+    # Arrow does not support direct casting from double to date32(), but for
+    # integer-like values we can go via int32()
+    # https://issues.apache.org/jira/browse/ARROW-15798
+    # TODO revisit if arrow decides to support double -> date casting
+    x <- build_expr("cast", x, options = cast_options(to_type = int32()))
+    delta_in_sec <- call_binding("difftime", origin, "1970-01-01")
+    delta_in_sec <- build_expr("cast", delta_in_sec, options = cast_options(to_type = int64()))
+    delta_in_days <- (delta_in_sec / 86400L)$cast(int32())
+    x <- build_expr("+", x, delta_in_days)
```
Comment (jonkeane) [inline r/R/dplyr-funcs-datetime.R:418]: I'm a little bit confused about this section. I suspect that we might be conflating a few things together which is making this hard(er) to read than it needs to be:

* For the condition, we trigger this block if `x` is numeric and either (or both) x is not an integer or the origin is non-default. And that means we are conflating our float work around with allowing a non-default origin. I bet if we separated these two out, that would make this clearer + easier to maintain. Feel free to break this out into a separate smaller help function if that's beneficial.
* The casting on `delta_in_sec` seems off — is there a reason you want|need to case that to int64, and then later divide by something that can fit in int32, then bast it to int32? Could you start off as int32 on line 416?

File: r/R/dplyr-funcs-datetime.R [hunk 8]
```diff
@@ -372,3 +372,50 @@ binding_format_datetime <- function(x, format = "", tz = "", usetz = FALSE) {
...
+  if (call_binding("is.Date", x)) {
+    return(x)
+
+    # cast from POSIXct
+  } else if (call_binding("is.POSIXct", x)) {
+    # base::as.Date() first converts to the desired timezone and then extracts
+    # the date, which is why we need to go through timestamp() first
+    if (base || !is.null(tz)) {
+      x <- build_expr("cast", x, options = cast_options(to_type = timestamp(timezone = tz)))
+    }
```
Comment (jonkeane) [inline r/R/dplyr-funcs-datetime.R:396]: Would it be possible to put this exceptional case in the binding for `as.Date` before then calling `binding_as_date`? I'm not 100% sure it would be cleaner, necessarily, but it would make this helper function much cleaner, and keep the exceptional part of the binding with the binding declaration bit.

File: r/tests/testthat/test-dplyr-funcs-type.R [hunk 9]
```diff
@@ -854,21 +874,94 @@ test_that("as.Date() converts successfully from date, timestamp, integer, char a
...
+  # TODO revisit after https://issues.apache.org/jira/browse/ARROW-15798
+  expect_error(
+    test_df %>%
+      arrow_table() %>%
       mutate(date_double = as.Date(double_var, origin = "1970-01-01")) %>%
+      collect()
+  )
+
+  # we do not support as_date with double/ float (error surfaced from C++)
+  # TODO revisit after https://issues.apache.org/jira/browse/ARROW-15798
```
Comment (jonkeane) [inline r/tests/testthat/test-dplyr-funcs-type.R:888]:
```suggestion
  # TODO: revisit after https://issues.apache.org/jira/browse/ARROW-15798
```

File: r/tests/testthat/test-dplyr-funcs-type.R [hunk 10]
```diff
@@ -854,21 +874,94 @@ test_that("as.Date() converts successfully from date, timestamp, integer, char a
...
+        ddouble_date = as_datetime(double_date)
+      ) %>%
       collect(),
-     test_df,
-     warning = TRUE
+    regexp = "Float value 10.1 was truncated converting to int64"
   )
 
+  # separate tz test so we can skip on Windows
   skip_on_os("windows") # https://issues.apache.org/jira/browse/ARROW-13168
```
Comment (jonkeane) [inline r/tests/testthat/test-dplyr-funcs-type.R:960]: We shouldn't need to skip on windows anymore, right?

**Review by jonkeane — CHANGES_REQUESTED** [review]
Thanks, a few comments + questions for you

File: r/R/dplyr-funcs-datetime.R (hunk 6 above)
Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R:399]: Done. Removed self-asignment.

File: r/tests/testthat/test-dplyr-funcs-type.R (hunk 9 above)
Comment (dragosmg) [reply r/tests/testthat/test-dplyr-funcs-type.R:888]: Done

File: r/tests/testthat/test-dplyr-funcs-type.R (hunk 10 above)
Comment (dragosmg) [reply r/tests/testthat/test-dplyr-funcs-type.R]: Right

File: r/R/dplyr-funcs-datetime.R (hunk 8 above)
Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R:396]: We run row 395 in 2 cases: either when we call `as.Date()` or when we call `as_date()` and the user passes a value for `tz`. So the row is run both by the `as.Date()` and `as_date()` bindings. It encapsulates the main difference between the 2 functions when it comes to the treatment of `tz` (either passed as argument or attribute of the `POSIXct` object)

File: r/R/dplyr-funcs-datetime.R (hunk 7 above)
Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R:418]: I think my thinking (for conflating numeric with origin) was that the origin `argument` only makes sense when `x` is numeric. If we had separate methods, then `origin` would only be an arg for the numeric method.

Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R]: Done. Both points addressed:
* conditions separated for `origin` different from UNIX epoch
* `delta_in_sec` is cast to `int32()` when created.

Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R]: I know re-discovered why the cast to `int64()` was needed. duration -> int32 is not supported, which means row 416 above errors.

Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R]: reverted.

Comment (jonkeane) [reply r/R/dplyr-funcs-datetime.R]: Is there a JIRA for that? We should flag it since it looks like kind of like a mistake rather than a work around to get something working.

File: r/R/dplyr-funcs-datetime.R (hunk 8 above)
Comment (jonkeane) [reply r/R/dplyr-funcs-datetime.R:487]: Aaah so maybe it would be helpful to call that argument something else? Like maybe `use_tz`? That way the intent of it is a bit clearer on first read? (And of course comments if they are helpful...)

File: r/tests/testthat/test-dplyr-funcs-type.R (hunk 10 above)
Comment (jonkeane) [reply r/tests/testthat/test-dplyr-funcs-type.R]: We don't have to do it here, but we should get rid of those ASAP so we see if there are any failures before the release

File: r/R/dplyr-funcs-datetime.R (hunk 7 above)
Comment (jonkeane) [reply r/R/dplyr-funcs-datetime.R]: > I think my thinking (for conflating numeric with origin) was that the origin argument only makes sense when x is numeric. If we had separate methods, then origin would only be an arg for the numeric method.

Right, that totally makes sense that origin is only relevant for when `x` is numeric. My comment was mostly about how hard it is to reason through that if/else clause to figure out that was what was doing on. I do really think this chunk of code would be easier to reason about if we broke it up and did each type conversion on its own. We don't really need or want to do full-on method dispatch, but if we broke up each piece as a separate function it'll make what's going on in each + why much easier to follow.

File: r/R/dplyr-funcs-datetime.R (hunk 6 above)
Comment (jonkeane) [reply r/R/dplyr-funcs-datetime.R]: Do we need this line at all, though? Now it's just `x` but that should be a no-op. What happens if we remove the line entirely?

File: r/tests/testthat/test-dplyr-funcs-type.R (hunk 10 above)
Comment (dragosmg) [reply r/tests/testthat/test-dplyr-funcs-type.R]: I had removed those relevant for this PR.

File: r/R/dplyr-funcs-datetime.R (hunk 7 above)
Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R]: Makes sense.

Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R]: On the Jira for the cast from `duration` to `int32`. There isn't one. I'm not sure how clearly documented this is, but it seems int64 is the matching type for quite a bit of the date-time/ duration types. We could extend the scope of [ARROW-15858](https://issues.apache.org/jira/browse/ARROW-15858)? Would that make sense?

File: r/R/dplyr-funcs-datetime.R (hunk 8 above)
Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R]: Done. Changed the name to `use_tz` and added, what I think are, clarifying comments.

File: r/R/dplyr-funcs-datetime.R (hunk 6 above)
Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R]: Removed the line. Local tests pass.

File: r/R/dplyr-funcs-datetime.R (hunk 7 above)
Comment (jonkeane) [reply r/R/dplyr-funcs-datetime.R]: > On the Jira for the cast from duration to int32. There isn't one. I'm not sure how clearly documented this is, but it seems int64 is the matching type for quite a bit of the date-time/ duration types. We could extend the scope of [ARROW-15858](https://issues.apache.org/jira/browse/ARROW-15858)? Would that make sense?

Right, it makes sense that we have the most coverage on int64, since that's what most of the date, time, duration types are stored as. But IMO there's no reason we shouldn't add to C++ int32 -> various types date, time duration types if int64 already works for them. 

I think ARROW-15858 is slightly separate cause it's dealing with objects of `difftime` in R (though it is related...).  

But this behavior is not great, in an ideal world, we shouldn't need to go from int32 -> int64 to then  go to a duration whenever we do these. A targeted Jira that describes that would be 💯 (and linked here so we know what we can get rid of when that's done)
```
> int32 <- Scalar$create(1, type=int32())
> int32$cast(duration())
Error: NotImplemented: Unsupported cast from int32 to duration using function cast_duration
> int64 <- Scalar$create(1, type=int64())
> int64$cast(duration())
Scalar
1
```

Tangentially: We might want to clean up the language on that ticket too, mentioning floats in there might be confusing since we already know that in Arrow going from float -> duration is not working (yet?):

```
> float <- Scalar$create(1, type=float())
> type(float)
Float32
float
> float$cast(duration())
Error: NotImplemented: Unsupported cast from float to duration using function cast_duration
```


> Makes sense. I created [ARROW-16257](https://issues.apache.org/jira/browse/ARROW-16257) for breaking it up in individual functions. I'd like to get this merged in time for 8.0.0 release and then we can follow-up to break each piece into an individual function. What do you think?

I won't demand that we do the refactoring now — but we do have to get the behavior right, and I suspect that refactoring it now will make clearer more easily what the issues are and how to fix them.

File: r/R/dplyr-funcs-datetime.R [hunk 11]
```diff
@@ -463,3 +463,54 @@ duration_from_chunks <- function(chunks) {
   }
   duration
 }
+
+binding_as_date <- function(x,
+                            format = NULL,
+                            tryFormats = "%Y-%m-%d",
+                            origin = "1970-01-01",
+                            use_tz = "UTC",
+                            base = TRUE) {
```
Comment (jonkeane) [inline r/R/dplyr-funcs-datetime.R:472]: Sorry if I was unclear on this: I meant that we should rename `base` to `use_tz` and keep `tz` like it is. Though even still, like I said the first time: I suspect we don't actually need this `base` argument at all, and instead we can alter `x` to include the TZ before calling `binding_as_date` and then not carry around this extra argument: 

basically, you make the `as.Date` binding be:

```
if (call_binding("is.POSIXct", x)) {
      x <- build_expr("cast", x, options = cast_options(to_type = timestamp(timezone = tz)))
}
binding_as_date(x, ...)
```

But if the argument is necessary, let's call it something that is explicit about what it does.

Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R:472]: I see what you mean now. I have a new variant in which `binding_as_date()` no longer needs a `tz` argument since `tz` is used outside it (inside the binding definition). I kinda like this approach as it brings front and center the difference between `base::as.Date()` and `lubridate::as_date()`.

Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R]: Done

File: r/R/dplyr-funcs-datetime.R (hunk 7 above)
Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R]: I refactored it.

Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R]: I'm not sure how much traction we would get for enabling `int32` -> `duration`. Judging by the comments on the [time32 and time64 to duration Jira](https://issues.apache.org/jira/browse/ARROW-15996), I'd say probably not much.

Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R]: There is also [ARROW-15862](https://issues.apache.org/jira/browse/ARROW-15862) touching on the issue of conversion from numeric (integers and floats) to duration.

File: r/R/dplyr-funcs-datetime.R [hunk 12]
```diff
@@ -463,3 +463,56 @@ duration_from_chunks <- function(chunks) {
...
+  # integer-like values we can go via int32()
+  # https://issues.apache.org/jira/browse/ARROW-15798
+  # TODO revisit if arrow decides to support double -> date casting
+  if (!call_binding("is.integer", x)) {
+    x <- build_expr("cast", x, options = cast_options(to_type = int32()))
+  }
+
+  if (origin != "1970-01-01") {
+    delta_in_sec <- call_binding("difftime", origin, "1970-01-01")
+    delta_in_sec <- build_expr("cast", delta_in_sec, options = cast_options(to_type = int64()))
```
Comment (jonkeane) [inline r/R/dplyr-funcs-datetime.R:512]: Would you mind adding the issue that I presume you created for int32 -> duration here?

Comment (jonkeane) [reply r/R/dplyr-funcs-datetime.R:512]: And also(?) or only(?) the issue you created for making the wrapper we talked about on https://github.com/apache/arrow/pull/12855

File: r/R/dplyr-funcs-datetime.R [hunk 13]
```diff
@@ -463,3 +463,56 @@ duration_from_chunks <- function(chunks) {
...
+    # cast from character
+  } else if (call_binding("is.character", x)) {
+    x <- binding_as_date_character(x, format, tryFormats)
+
+    # cast from numeric
+  } else if (call_binding("is.numeric", x)) {
+    x <- binding_as_date_numeric(x, origin)
+  }
+
+  build_expr("cast", x, options = cast_options(to_type = date32()))
```
Comment (jonkeane) [inline r/R/dplyr-funcs-datetime.R:488]: This + the helper functions below are fantastic! Makes it so much easier to isolate "which of this is about the type?" and "which of this is about the vagaries of `as.Date`?".

File: r/tests/testthat/test-dplyr-funcs-type.R [hunk 14]
```diff
@@ -845,6 +854,15 @@ test_that("as.Date() converts successfully from date, timestamp, integer, char a
...
   )
 
+  # strptime does not support a partial format -
+  # TODO revisit once - https://issues.apache.org/jira/browse/ARROW-15813
+  expect_error(
+    test_df %>%
+      arrow_table() %>%
+      mutate(date_char_ymd = as_date(character_ymd_var)) %>%
+      collect()
+  )
```
Comment (jonkeane) [inline r/tests/testthat/test-dplyr-funcs-type.R:864]: Is the error here coming from C++ only (and is that why you're not asserting what the error is)? If yes to both: can we add something to the comment. We should be clear (to our future selves) why we are going against our pattern that one should always assert what the error is when using `expect_error`.

File: r/R/dplyr-funcs-datetime.R [hunk 15]
```diff
@@ -463,3 +463,56 @@ duration_from_chunks <- function(chunks) {
   }
   duration
 }
+
+binding_as_date <- function(x,
+                            format = NULL,
+                            tryFormats = "%Y-%m-%d",
+                            origin = "1970-01-01") {
```
Comment (jonkeane) [inline r/R/dplyr-funcs-datetime.R:470]: I don't think we should change this here now, but something to think about and maybe do in the future if we have similar setups like this: It would probably be easier to pass the args down using `...` that way we don't always have to write out all of the arguments if we were to add one. 

Like I said, let's not change this now, but wanted to suggest that.

File: r/R/dplyr-funcs-datetime.R [hunk 16]
```diff
@@ -463,3 +463,56 @@ duration_from_chunks <- function(chunks) {
...
+binding_as_date_character <- function(x,
+                                      format = NULL,
+                                      tryFormats = "%Y-%m-%d") {
+  format <- format %||% tryFormats[[1]]
+  # unit = 0L is the identifier for seconds in valid_time32_units
+  build_expr("strptime", x, options = list(format = format, unit = 0L))
+}
+
+binding_as_date_numeric <- function(x,
+                                  origin = "1970-01-01") {
```
Comment (jonkeane) [inline r/R/dplyr-funcs-datetime.R:500]:
```suggestion
binding_as_date_numeric <- function(x, origin = "1970-01-01") {
```

Minor style nitpick, this could all to on one line, it's quite short still.

**Review by jonkeane — CHANGES_REQUESTED** [review]
This is so so so close! A few minor style things, and a few requests for some comments.

File: r/R/dplyr-funcs-datetime.R (hunk 12 above)
Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R:512]: Issues added as comments, but will also add them here:
* [ARROW-16253](https://issues.apache.org/jira/browse/ARROW-16253) - wrapper function for the casting chain `double` -> `int64` -> `duration`
* [ARROW-15862](https://issues.apache.org/jira/browse/ARROW-15862) - support in C++ casting from `int32` or `double` to `duration`

File: r/R/dplyr-funcs-datetime.R (hunk 13 above)
Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R:488]: 💯

File: r/tests/testthat/test-dplyr-funcs-type.R (hunk 14 above)
Comment (dragosmg) [reply r/tests/testthat/test-dplyr-funcs-type.R:864]: Done

File: r/R/dplyr-funcs-datetime.R (hunk 16 above)
Comment (dragosmg) [reply r/R/dplyr-funcs-datetime.R:500]: Sure. Done

**Comment by dragosmg** [comment]
I think the CI fail is unrelated to the changes in the PR

**Comment by jonkeane** [comment]
> I think the CI fails are unrelated to the changes in the PR

It's good practice to enumerate which are failing + link to the issues that are resolving them, or that they are failing on master, etc.

Both the R / AMD64 Ubuntu 20.04 R 4.1 build and the linter should be resolved by https://github.com/apache/arrow/pull/12945 or https://github.com/apache/arrow/pull/12958
